### ISSUE TITLE
Gaussianblur Performance Difference between OpenCV 3.3.0 and OpenCV 4.7.0

### ISSUE THREAD

**Issue by christian1steinmetz** [opencv/opencv#24135]
### System Information

OpenCV version: 4.7.0 and 3.3.0
Operating System / Platform: Windows 10
Compiler & compiler version: Visual Studio 2017

### Detailed description

I recently upgraded an older software project that used to run using OpenCV 3.3.0 to the version of OpenCV that was then up-to-date (4.7.0). For some reason, Somewhere inside the project, a grayscale image is blurred using the function GaussianBlur. Even though the input images and all parameters are similar, the output is very different between the output that is obtained from the 3.3.0-version and the 4.7.0-version.
I call the function as follows:

```
#include <opencv2/opencv.hpp>

int main()
{
cv::Mat img = cv::imread("<path>", IMREAD_GRAYSCALE);
cv::Mat blurred;
cv::resize(img , img , { 160, 90 });
cv::GaussianBlur(frame, blurred, {0, 0}, 7.5);
}
```

The result is different in almost every pixel, by usually less than 10, but always around 3-5 intensity steps. I tried to debug into OpenCV and found a completely new implementation using the OpenCV parallel_for_ in the newer version in contrast to a simpler hal::sepFilter2D in OCV 3.3.0, so it was hard to compare intermediate steps. 

It seems highly suspicious and very unlogical for me, though, that the results for a fixed algorithm are this different between the two versions. I did not find anything mentioned in any of the release notes up to now, stating that the old version of the Gaussian blur was wrong to some degree, but this leaves me very confused.

Does anyone have any insights? Imho, a software update should not change the quality of standard algorithms, at least not if it's not purely concerning speed.

[This is a question, but related to critical insights in the code, so I figured this is the better place to put it than in the forum where users are probably more present than contributors.]

### Steps to reproduce

Generate two projects, one with OpenCV 3.3.0 and one with OpenCV 4.7.0 and use the same main-file (above). Look at the differences. e.g. with the help of Image Watch.

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [ ] I updated to the latest OpenCV version and the issue is still there
- [ ] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by LaurentBerger** [comment]
What's your image format (jpg, png, ...)?

**Comment by zihaomu** [comment]
Hi @christian1steinmetz, I try to reproduce your issue on my Mac x86, python.
And I found it always generates the exact same image both for jpeg and png, test with `4.8`, `4.7`, `3.4` versions.
I will try it on my Windows platform later.

**Comment by opencv-alalek** [comment]
https://github.com/opencv/opencv/wiki/OE-15.-Bit-Exactness

Check #10345 (since 3.4.1) and several other fixes.

**Comment by christian1steinmetz** [comment]
> What's your image format (jpg, png, ...)?

This does not really matter. I loaded both a jpg file and a frame from a video, but did check the cv:Mat that goes into GaussianBlur. It is the same, the output is not, though

**Comment by christian1steinmetz** [comment]
> Hi @christian1steinmetz, I try to reproduce your issue on my Mac x86, python. And I found it always generates the exact same image both for jpeg and png, color and gray, test with `4.8`, `4.7`, `3.4`, and `3.3` versions.
> 
> Test on Windows i7-9750 platform, opencv-python `3.4.18.65` and `4.7.0.72` It always generates the same image.

Hmm, this seems extremely weird as the only difference comes directly from the OpenCV-function GaussianBlur, other operations are bit-perfect. And I did not change any hardware in between the tests.

**Comment by christian1steinmetz** [comment]
As already indicated in the pictures, I started to check which version makes most sense for me by looking at the means of the images (Gaussian blur should not change the mean of images (apart from quantization, of course), at least not anywhere apart from the borders).

I found these results:
                          |                 frame             |         blurred
OpenCV 3.3.0    |    138.78569444444443 | 143.30013888888888
OpenCV 4.7.0    |    138.78569444444443 | 138.92506944444443

This makes me believe that the newer version is indeed more correct than the old one. Maybe I missed something in the release notes (or this actually is an issue that only arises in some very specific environments)

**Comment by zihaomu** [comment]
> This makes me believe that the newer version is indeed more correct than the old one.

Agree.

**Comment by LaurentBerger** [comment]
> This does not really matter. I loaded both a jpg file and a frame from a video, but did check the cv:Mat that goes into GaussianBlur. It is the same, the output is not, though

You are wrong. It's really matter libjpeg or libturbo-jpeg do not give same image. If you want to compare result you have to use png format or better save image in yml format and read data in yml format

**Comment by christian1steinmetz** [comment]
> > This does not really matter. I loaded both a jpg file and a frame from a video, but did check the cv:Mat that goes into GaussianBlur. It is the same, the output is not, though
> 
> You are wrong. It's really matter libjpeg or libturbo-jpeg do not give same image. If you want to compare result you have to use png format or better save image in yml format and read data in yml format

I mean, yes of course, it does matter when loading an image. It could lead to differences between the versions when loading the image with the different versions iff another library is used for loading jpgs, i..e integer accuracy is not achieved during decompression. 
But as I did compare the images directly before and after the GaussianBlur (I even saved the images at the resprective steps as binary files and loaded them and computed a diff which is zero before and non-zero [for every kernel size>9] afterwards), this does definitely not matter in this context.
